client/app/pages/dashboards/public-dashboard-page.js

```diff
@@ -25,40 +25,44 @@ const PublicDashboardPage = {
 
     this.logoUrl = logoUrl;
     this.public = true;
-    this.dashboard.widgets = Dashboard.prepareDashboardWidgets(this.dashboard.widgets);
+    this.globalParameters = [];
 
-    const refreshRate = Math.max(30, parseFloat($location.search().refresh));
-
-    if (refreshRate) {
-      const refresh = () => {
-        loadDashboard($http, $route).then((data) => {
-          this.dashboard = data;
-          this.dashboard.widgets = Dashboard.prepareDashboardWidgets(this.dashboard.widgets);
+    this.extractGlobalParameters = () => {
+      this.globalParameters = this.dashboard.getParametersDefs();
+    };
 
-          this.filters = []; // TODO: implement (@/services/dashboard.js:collectDashboardFilters)
-          this.filtersOnChange = (allFilters) => {
-            this.filters = allFilters;
-            $scope.$applyAsync();
-          };
+    $scope.$on('dashboard.update-parameters', () => {
+      this.extractGlobalParameters();
+    });
 
-          $timeout(refresh, refreshRate * 1000.0);
-        });
-      };
+    const refreshRate = Math.max(30, parseFloat($location.search().refresh));
 
-      $timeout(refresh, refreshRate * 1000.0);
-    }
+    this.refreshDashboard = () => {
+      loadDashboard($http, $route).then((data) => {
+        this.dashboard = new Dashboard(data);
+        this.dashboard.widgets = Dashboard.prepareDashboardWidgets(this.dashboard.widgets);
+        this.dashboard.widgets.forEach(widget => widget.load());
+        this.filters = []; // TODO: implement (@/services/dashboard.js:collectDashboardFilters)
+        this.filtersOnChange = (allFilters) => {
+          this.filters = allFilters;
+          $scope.$applyAsync();
+        };
+
+        this.extractGlobalParameters();
+      });
+
+      if (refreshRate) {
+        $timeout(this.refreshDashboard, refreshRate * 1000.0);
```

Comment:
I tried, This does not work.

Reply:
@arikfr @truebit could you elaborate on why you think this does not work? Because when I'm adding a `refresh=30` querystring parameter, I'm seeing public dashboards update with new results just fine.

Reply:
@rauchy the queries you were testing with were scheduled?

Reply:
@arikfr no they weren't. Why? Are public dashboard refreshes not supposed to trigger new query execution?

Reply:
@rauchy `refreshDashboard` calls `loadDashboard`, which just loads the dashboard definition. How is it supposed to trigger a refresh this way?

Reply:
@arikfr this seems like a (desired?) side-effect of [widget.load()](https://github.com/getredash/redash/blob/master/client/app/pages/dashboards/public-dashboard-page.js#L37).

Reply:
The logic there is a bit hard to follow, but if I'm not mistaken it will bring the last result and not trigger new execution (which is what the user expects in case of a refresh).

Reply:
Yeah, you're right. Public dashboards don't refresh data automatically, but they do update with last results according to the `refresh` querystring parameter. I think the reason I thought they do is that I had the same dashboard open in a standard (non-shared) view, and it was triggering refreshes (with `maxAge=0`).

So regarding

> We probably need to switch to same refresh as the regular dashboard?

Should I add this to public dashboards? (with same behavior as regular dashboards)

![image](https://user-images.githubusercontent.com/289488/60879996-6909af00-a24b-11e9-9e6c-2ecc60a70136.png)

Reply:
> Should I add this to public dashboards? (with same behavior as regular dashboards)

No, for now let's not add more features but focus on finishing this. So just need to make it work, but no UI is needed. Later we can introduce the UI.

Reply:
Although now I wonder if all the changes make it possible now to reuse the same dashboard page for public dashboards and reduce code duplication?

Reply:
> Although now I wonder if all the changes make it possible now to reuse the same dashboard page for public dashboards and reduce code duplication?

I feel that the regular dashboard page is still an order of magnitude more complex than the shared dashboard page, i.e. it has a lot more concerns - saving, sharing, archiving, layout etc), while the shared dashboard page is a simpler, read-only version.

Definitely something to think about for future refactoring, but I think we can safely live with this minimal duplication, as it feels less costly than a grand refactoring ATM.

> No, for now let's not add more features but focus on finishing this. So just need to make it work, but no UI is needed. Later we can introduce the UI.

So just allow data refreshes using the `refresh` querystring parameter?

Reply:
> So just allow data refreshes using the refresh querystring parameter?

Yes.

Reply:
500c155